Fix computed value for `marker` (SVG)

rwlbuis opened the pull request:
#### a04206ec3b049480111fdc1d6f791202e261f1bd
<pre>
Fix computed value for `marker` (SVG)
<a href="https://bugs.webkit.org/show_bug.cgi?id=294206">https://bugs.webkit.org/show_bug.cgi?id=294206</a>
<a href="https://rdar.apple.com/153373245">rdar://153373245</a>

Reviewed by Sam Weinig.

Align `marker` computed value with the specification [1].

[1] <a href="https://svgwg.org/svg2-draft/painting.html#MarkerShorthand">https://svgwg.org/svg2-draft/painting.html#MarkerShorthand</a>

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/marker-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/marker-computed.svg:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/painting/parsing/marker-computed-expected.txt: Removed.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::ExtractorCustom::extractMarkerShorthandSerialization):
(WebCore::Style::ExtractorCustom::extractMarkerShorthand):

Canonical link: <a href="https://commits.webkit.org/301462@main">https://commits.webkit.org/301462@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/10acd14617a52a855d32d0104c371a9af94d970b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45553 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36308 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132761 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77756 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127762 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46236 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54111 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95909 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64012 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36973 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112577 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76400 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35888 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30761 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76233 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106755 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30974 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135444 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52676 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40411 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104379 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53125 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108787 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104106 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26551 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49477 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27795 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50047 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52570 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58382 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51914 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55265 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53612 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Ahmad-S792 commented on LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/painting/parsing/marker-computed-expected.txt:
We can remove this platform specific expectation now.

rwlbuis replied:
Good point, removed.

weinig commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -3136,5 +3138,31 @@ inline void ExtractorCustom::extractOrphansSerialization(ExtractorState& state,
     extractSerialization<CSSPropertyOrphans>(state, builder, context);
 }
 
+inline void ExtractorCustom::extractMarkerShorthandSerialization(ExtractorState& state, StringBuilder& builder, const CSS::SerializationContext& context)
```
I don't believe this is correct.

When the three longhands are not the same, it should serialize as the empty string (or return a nullptr for RefPtr<CSSValue>), not `none`. (see https://github.com/w3c/svgwg/issues/983)

Something like:

```
inline RefPtr<CSSValue> ExtractorCustom::extractMarkerShorthand(ExtractorState& state)
{
    auto markerStart = state.style.markerStart();
    auto markerMid = state.style.markerMid();
    auto markerEnd = state.style.markerEnd();
    if (markerStart == markerMid && markerMid == markerEnd)
        return createCSSValue(state.pool, state.style, *markerStart);
    return nullptr;
}

inline void ExtractorCustom::extractMarkerShorthandSerialization(ExtractorState& state, StringBuilder& builder, const CSS::SerializationContext& context)
{
    auto markerStart = state.style.markerStart();
    auto markerMid = state.style.markerMid();
    auto markerEnd = state.style.markerEnd();
    if (markerStart == markerMid && markerMid == markerEnd)
        serializationForCSS(builder, context, state.style, *markerStart);
}
```

If there aren't any tests for the different values for different markers case, we should add those to WPT.

weinig commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -3135,5 +3137,31 @@ inline void ExtractorCustom::extractOrphansSerialization(ExtractorState& state,
+    auto markerMid = state.style.markerMid();
+    auto markerEnd = state.style.markerEnd();
+    if (markerStart == markerMid && markerMid == markerEnd) {
+        if (auto url = markerStart.tryURL())
```
Is this if statement needed? The market should be directly serializable/convertible to CSSValue?

rwlbuis replied:
I thought I tried it and ran into build problems when I tried it before, but I retried and indeed that works just fine, fixed.

weinig replied:
They still seem to be there for the serialization function. Also, you want the to use the non-CSS:: prefixed serializationForCSS, which also takes a RenderStyle here.

rwlbuis replied:
Ah, good point, done.

rwlbuis committed:
Fix computed value for `marker` (SVG)
https://bugs.webkit.org/show_bug.cgi?id=294206
rdar://153373245

Reviewed by Sam Weinig.

Align `marker` computed value with the specification [1].

[1] https://svgwg.org/svg2-draft/painting.html#MarkerShorthand

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/marker-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/marker-computed.svg:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/painting/parsing/marker-computed-expected.txt: Removed.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::ExtractorCustom::extractMarkerShorthandSerialization):
(WebCore::Style::ExtractorCustom::extractMarkerShorthand):

Canonical link: https://commits.webkit.org/301462@main
